src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala

```diff
@@ -258,15 +330,56 @@ case class YulObject(name: String, code: Code, subobjects: Seq[YulObject], data:
             }
         }
 
-
-
         def dispatchCase(): codegen.Switch = codegen.Switch(Identifier("selector"), dispatchArray.toSeq)
 
-        val datasize: Expression = unary_ap("datasize", stringlit(runtimeObject))
-
-        def codeCopy() : Expression = triple("codecopy",ilit(0), unary_ap("dataoffset",stringlit(runtimeObject)), datasize)
+        val datasize: Expression = ap("datasize", stringlit(runtimeObject))
+
+        def codeCopy(): Expression = ap("codecopy", ilit(0), ap("dataoffset", stringlit(runtimeObject)), datasize)
+
+        def defaultReturn(): Expression = ap("return", ilit(0), datasize)
+
+        /**
+          * compute the abi tuple encode function for a given number of returns. for example, for 0,
+          * 1, and 2 returns:
+          *
+          * {{{
+          * function abi_encode_tuple__to__fromStack(headStart ) -> tail {
+          *    tail := add(headStart, 0)
+          * }
+          * }}}
+          *
+          * {{{
+          * function abi_encode_tuple_t_uint256__to_t_uint256__fromStack(headStart , value0) -> tail {
+          *    tail := add(headStart, 32)
+          *    abi_encode_t_uint256_to_t_uint256_fromStack(value0,  add(headStart, 0))
+          * }
+          * }}}
+          *
+          * {{{
+          * function abi_encode_tuple_t_uint256_t_uint256__to_t_uint256_t_uint256__fromStack(headStart , value0, value1) -> tail {
+          *    tail := add(headStart, 64)
+          *    abi_encode_t_uint256_to_t_uint256_fromStack(value0,  add(headStart, 0))
+          *    abi_encode_t_uint256_to_t_uint256_fromStack(value1,  add(headStart, 32))
+          * }
+          * }}}
+          *
+          * @param n the number of returns
+          * @return the function definition for output
+          */
+        def write_abi_encode(n: Int): FunctionDefinition = {
+            val var_indices: Seq[Int] = Seq.tabulate(n)(i => i)
+            val encode_lines: Seq[YulStatement] = var_indices.map(i =>
+                ExpressionStatement(ap("abi_encode_t_uint256_to_t_uint256_fromStack",
+                    Identifier("value" + i.toString),
+                    ap("add", Identifier("headStart"), ilit(n * 32)))))
+
+            val bod: Seq[YulStatement] = assign1(Identifier("tail"), ap("add", Identifier("headStart"), ilit(32 * n))) +: encode_lines
+            FunctionDefinition("abi_encode_tuple_to_fromStack" + n.toString,
+                TypedName("headStart", "") +: var_indices.map(i => TypedName("value" + i.toString, "")),
+                Seq(TypedName("tail", "")), Block(bod))
+        }
 
-        def defaultReturn(): Expression = binary_ap("return", ilit(0), datasize)
+        def abiEncodeTupleFuncs(): YulStatement = Block(abiEncodesNeeded.map(write_abi_encode).toSeq)
```

Comment:
why abiEncodesNeeded.map(write_abi_encode) rather than write_abi_encode(abiEncodesNeeded)?

Reply:
`abiEncodesNeeded` is a set of integers but `write_abi_encode` only takes one integer, so you have to iterate over the collection to get the encoding functions you need in a given run.